### ISSUE TITLE
Tracking Issue for Explicit Tail Calls

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust#112788]
This is a tracking issue for the RFC "Explicit Tail Calls" (rust-lang/rfcs#3407) (note that RFC is not accepted yet, this feature is currently implemented as part of an experiment).
The feature gate for the issue is `#![feature(explicit_tail_calls)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Implement the RFC
    - [x] Parsing/HIR/MIR
    - [x] Implement the necessary checks for tail calls
    - [x] Add support in LLVM backend (draft commits: https://github.com/rust-lang/rust/pull/112657/commits/dee279d5a81f3ad2e07a90365da49ff7e382feeb, https://github.com/rust-lang/rust/pull/112657/commits/0722a1ec41a4f903707ca2b75d6644f89ac2a1f9)
       - [ ] Handle indirect ABI arguments
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

- What exactly are the semantics of passing indirect arguments? What about by-move arguments? Also see https://github.com/rust-lang/rust/issues/144820.

(fill when the rfc is merged)

### Implementation history
- [x] https://github.com/rust-lang/rust/pull/112790
- [x] https://github.com/rust-lang/rust/pull/112791
- [x] https://github.com/rust-lang/rust/pull/112887
- [x] https://github.com/rust-lang/rust/pull/113093
- [ ] https://github.com/rust-lang/rust/pull/113095
- [x] https://github.com/rust-lang/rust/pull/113128
- [x] https://github.com/rust-lang/rust/pull/133607
- [x] https://github.com/rust-lang/rust/pull/135973
- [x] https://github.com/rust-lang/rust/pull/135976
- [x] #144232 
- [x] https://github.com/rust-lang/rust/pull/144756
- [x] https://github.com/rust-lang/rust/pull/144817
- [x] https://github.com/rust-lang/rust/pull/144835
- [x] https://github.com/rust-lang/rust/pull/144851
- [x] https://github.com/rust-lang/rust/pull/144865
- [x] https://github.com/rust-lang/rust/pull/144917
- [x] https://github.com/rust-lang/rust/pull/144920
- [ ] https://github.com/rust-lang/rust/pull/144933
- [x] https://github.com/rust-lang/rust/pull/144915

**Comment by scottmcm** [comment]
While looking at something else, I tried using `tailcc`, and got sad:
```
LLVM ERROR: Can't handle guaranteed tail call under win64 yet
error: could not compile `core` (lib)
``

**Comment by travis-leith** [comment]
@scottmcm can you clarify: is this issue (https://github.com/llvm/llvm-project/issues/53271) holding up implemention of explicit tail calls in rust?

**Comment by WaffleLapkin** [comment]
@travis-leith this is not technically blocking the implementation, to my knowledge. But it is a big/noticeable problem for testing & stabilization.

**Comment by xacrimon** [comment]
https://github.com/llvm/llvm-project/issues/53271 seems fixed now on clang 20.10, so this might be able to move forward?

**Comment by zdivelbiss** [comment]
Just curious, what the reason for naming this `become`? I'm not one for often bikeshedding, but this incredibly general and perhaps intentionally un-descriptive/un-explicit.

Is the naming of the keyword temporary?

**Comment by oxalica** [comment]
> Just curious, what the reason for naming this `become`? I'm not one for often bikeshedding, but this incredibly general and perhaps intentionally un-descriptive/un-explicit.
> 
> Is the naming of the keyword temporary?

See the RFC issue for more discussion about it, like https://github.com/rust-lang/rfcs/pull/3407#issuecomment-3042659375. 
> To me the become keyword makes a lot of sense if you think about it as follows: make the current function (frame) become this new one. That is also what is kind of happening under the hood. As a minor detail: become is already a keyword, reserved since 2014, by Graydon, for exactly this purpose.

So the original purpose of reserving `become` as a keyword (since 2014) is exactly for tail calls. Then using it sounds the most intuitive[^1] choice, and it is not temporary.

[^1]: Personally speaking, "become" makes more sense than "tail call". What does "tail" mean? "X call" sounds like a "call" with extra steps, while tail call is actually parallel to a normal call (it's not `call` vs `call; ret`, it's `call` vs `jmp`).

**Comment by zdivelbiss** [comment]
Thank you for linking directly to the RFC; I somehow missed the existing discussion on the keyword (not used to mobile GitHub, perhaps).

I think `tailcall` is an existing and well-understood concept, however the keyword being already reserved is enough to satisfy.

Thanks again for your timely reply.

**Comment by phi-go** [comment]
I think it should also be mentioned that `become` is not set in stone, the language team (Josh) actually advised us not to use `become` for the experiment.

https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1579335011
> @phi-go Clarifying something: I am not at this time raising an objection to become as final syntax here. I'm objecting to using syntax in a nightly experiment that's one of the potential candidate syntaxes for the final version, on the basis that we'll then have a hard time evaluating syntax preferences and feedback independently from what people are used to on nightly. See @ssokolow's link to "[Don't Make the Demo look Done](https://headrush.typepad.com/creating_passionate_users/2006/12/dont_make_the_d.html)".
> 
> Could we green-light the experiment with some obvious non-final placeholder (e.g. something in a # namespace like k#), and continue the evaluation of syntax in parallel on the assumption that it can be easily changed?

Though, @WaffleLapkin also wrote a rebuttal at the time
https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1580758537

**Comment by coastalwhite** [comment]
Maybe as a nice case study of what this feature (and the "rust-preserve-none" feature) can bring. This is a benchmark of a specialized bytecode with and without those two features.

```
Benchmark 1 (9 runs): ./no-tailcall
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          3.69s  ± 23.0ms    3.67s  … 3.75s           1 (11%)        0%
  peak_rss           7.50MB ±    0      7.50MB … 7.50MB          0 ( 0%)        0%
  cpu_cycles         18.5G  ± 44.8M     18.5G  … 18.6G           0 ( 0%)        0%
  instructions       57.1G  ±  384K     57.1G  … 57.1G           0 ( 0%)        0%
  cache_references   1.93M  ±  539K     1.46M  … 3.30M           1 (11%)        0%
  cache_misses        126K  ± 69.8K     81.6K  …  308K           1 (11%)        0%
  branch_misses      95.0M  ±  672K     93.8M  … 96.1M           0 ( 0%)        0%
Benchmark 2 (9 runs): ./tailcall
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          3.40s  ± 15.8ms    3.38s  … 3.42s           0 ( 0%)        ⚡-  7.9% ±  0.5%
  peak_rss           7.55MB ± 10.9KB    7.52MB … 7.55MB          1 (11%)          +  0.7% ±  0.1%
  cpu_cycles         17.1G  ± 57.4M     17.0G  … 17.2G           0 ( 0%)        ⚡-  7.9% ±  0.3%
  instructions       58.0G  ± 63.2K     58.0G  … 58.0G           0 ( 0%)        💩+  1.6% ±  0.0%
  cache_references   1.77M  ±  330K     1.37M  … 2.35M           0 ( 0%)          -  8.2% ± 23.2%
  cache_misses        101K  ± 14.4K     85.1K  …  133K           0 ( 0%)          - 20.1% ± 40.0%
  branch_misses      50.0M  ±  500K     49.1M  … 50.6M           0 ( 0%)        ⚡- 47.3% ±  0.6%
```

Here I literally just replaced a loop with the `become` and `extern "rust-preserve-none"`. I think 8% runtime gains for such a small change are superb.

**Comment by Robbepop** [comment]
@coastalwhite very interesting numbers indeed! Thanks for sharing. :)

Let me follow-up with a benchmark for the Wasmi interpreter which has 4 different instruction dispatch modes where everything else is the same:

| Wasmi Config | CoreMark Score | Delta % |
|:---|:---:|:--:|
| direct tail-call | 2820 | |
| indirect tail-call | 2357 | -17% |
| loop-call | 1133 | -60% |
| loop-match | 1764 | -38% |

Higher score is better.

Source: https://github.com/wasmi-labs/wasm-coremark-rs

In conclusion: tail calls are necessary for great interpreter performance.
